broker/src/main/java/io/camunda/zeebe/broker/partitioning/topology/DynamicClusterTopologyService.java

```diff
@@ -0,0 +1,117 @@
+    final CompletableActorFuture<Void> started = new CompletableActorFuture<>();
+
+    final PartitionDistributionResolver partitionDistributionResolver =
+        new PartitionDistributionResolver();
+
+    clusterTopologyManagerService = getClusterTopologyManagerService(brokerStartupContext);
+
+    final var topologyManagerStartedFuture =
+        startClusterTopologyManager(
+            brokerStartupContext, clusterTopologyManagerService, partitionDistributionResolver);
+
+    topologyManagerStartedFuture.onComplete(
+        (ignore, topologyManagerFailed) -> {
+          if (topologyManagerFailed != null) {
+            started.completeExceptionally(topologyManagerFailed);
+          } else {
+            clusterTopologyManagerService
+                .getClusterTopology()
+                .onComplete(
+                    (topology, error) -> {
+                      if (error != null) {
+                        started.completeExceptionally(error);
+                      } else {
+                        try {
+                          partitionDistribution =
+                              partitionDistributionResolver.resolvePartitionDistribution(topology);
+                          started.complete(null);
+                        } catch (final Exception topologyConversionFailed) {
+                          started.completeExceptionally(topologyConversionFailed);
+                        }
+                      }
+                    });
+          }
+        });
+    return started;
+  }
+
+  @Override
+  public ActorFuture<Void> closeAsync() {
+    partitionDistribution = null;
+    if (clusterTopologyManagerService != null) {
+      return clusterTopologyManagerService.closeAsync();
+    } else {
+      return CompletableActorFuture.completed(null);
+    }
+  }
+
+  private static ActorFuture<Void> startClusterTopologyManager(
+      final BrokerStartupContext brokerStartupContext,
+      final ClusterTopologyManagerService clusterTopologyManagerService,
+      final PartitionDistributionResolver partitionDistributionResolver) {
+    final BrokerCfg brokerConfiguration = brokerStartupContext.getBrokerConfiguration();
+    final var allMembers =
+        PartitionDistributionResolver.getRaftGroupMembers(brokerConfiguration.getCluster());
+    final var localMemberId =
+        brokerStartupContext.getClusterServices().getMembershipService().getLocalMember().id();
+
+    final var otherMembers = allMembers.stream().filter(id -> !id.equals(localMemberId)).toList();
+
+    return clusterTopologyManagerService.start(
+        brokerStartupContext.getActorSchedulingService(),
+        otherMembers,
+        () ->
+            partitionDistributionResolver
+                .resolvePartitionDistribution(
+                    brokerConfiguration.getExperimental().getPartitioning(),
+                    brokerConfiguration.getCluster())
+                .partitions());
```

Comment:
I wanted to play around with this code so I came up with a patch that implements my suggestion. Still, feel free to ignore this.

<details><summary>Patch</summary>

```
Subject: [PATCH] refactor(topology): clarify what static configuration is used for
---
Index: broker/src/main/java/io/camunda/zeebe/broker/partitioning/topology/DynamicClusterTopologyService.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/broker/src/main/java/io/camunda/zeebe/broker/partitioning/topology/DynamicClusterTopologyService.java b/broker/src/main/java/io/camunda/zeebe/broker/partitioning/topology/DynamicClusterTopologyService.java
--- a/broker/src/main/java/io/camunda/zeebe/broker/partitioning/topology/DynamicClusterTopologyService.java	(revision 6f1ecfadd20bc27f34be1ecfdcd9612cc8392043)
+++ b/broker/src/main/java/io/camunda/zeebe/broker/partitioning/topology/DynamicClusterTopologyService.java	(revision 1180e012dfcc8567edf6b6f89d21f46db0c18aae)
@@ -8,13 +8,14 @@
 package io.camunda.zeebe.broker.partitioning.topology;
 
 import io.camunda.zeebe.broker.bootstrap.BrokerStartupContext;
-import io.camunda.zeebe.broker.system.configuration.BrokerCfg;
 import io.camunda.zeebe.scheduler.future.ActorFuture;
 import io.camunda.zeebe.scheduler.future.CompletableActorFuture;
 import io.camunda.zeebe.topology.ClusterTopologyManagerService;
+import io.camunda.zeebe.topology.TopologyInitializer.StaticConfiguration;
 import io.camunda.zeebe.topology.gossip.ClusterTopologyGossiperConfig;
 import java.nio.file.Path;
 import java.time.Duration;
+import java.util.stream.Collectors;
 
 public class DynamicClusterTopologyService implements ClusterTopologyService {
 
@@ -30,15 +31,10 @@
   @Override
   public ActorFuture<Void> start(final BrokerStartupContext brokerStartupContext) {
     final CompletableActorFuture<Void> started = new CompletableActorFuture<>();
-
-    final PartitionDistributionResolver partitionDistributionResolver =
-        new PartitionDistributionResolver();
-
     clusterTopologyManagerService = getClusterTopologyManagerService(brokerStartupContext);
 
     final var topologyManagerStartedFuture =
-        startClusterTopologyManager(
-            brokerStartupContext, clusterTopologyManagerService, partitionDistributionResolver);
+        startClusterTopologyManager(brokerStartupContext, clusterTopologyManagerService);
 
     topologyManagerStartedFuture.onComplete(
         (ignore, topologyManagerFailed) -> {
@@ -54,7 +50,7 @@
                       } else {
                         try {
                           partitionDistribution =
-                              partitionDistributionResolver.resolvePartitionDistribution(topology);
+                              PartitionDistributionResolver.resolvePartitionDistribution(topology);
                           started.complete(null);
                         } catch (final Exception topologyConversionFailed) {
                           started.completeExceptionally(topologyConversionFailed);
@@ -78,28 +74,27 @@
 
   private static ActorFuture<Void> startClusterTopologyManager(
       final BrokerStartupContext brokerStartupContext,
-      final ClusterTopologyManagerService clusterTopologyManagerService,
-      final PartitionDistributionResolver partitionDistributionResolver) {
-    final BrokerCfg brokerConfiguration = brokerStartupContext.getBrokerConfiguration();
-    final var allMembers =
-        PartitionDistributionResolver.getRaftGroupMembers(brokerConfiguration.getCluster());
+      final ClusterTopologyManagerService clusterTopologyManagerService) {
+    final var clusterCfg = brokerStartupContext.getBrokerConfiguration().getCluster();
+    final var partitionCfg =
+        brokerStartupContext.getBrokerConfiguration().getExperimental().getPartitioning();
+
+    final var partitionDistribution =
+        PartitionDistributionResolver.resolvePartitionDistribution(partitionCfg, clusterCfg)
+            .partitions();
+    final var allMembers = PartitionDistributionResolver.getRaftGroupMembers(clusterCfg);
     final var localMemberId =
         brokerStartupContext.getClusterServices().getMembershipService().getLocalMember().id();
 
-    final var otherMembers = allMembers.stream().filter(id -> !id.equals(localMemberId)).toList();
+    final var otherMembers =
+        allMembers.stream().filter(id -> !id.equals(localMemberId)).collect(Collectors.toSet());
+    final var staticConfiguration = new StaticConfiguration(partitionDistribution, otherMembers);
 
     return clusterTopologyManagerService.start(
-        brokerStartupContext.getActorSchedulingService(),
-        otherMembers,
-        () ->
-            partitionDistributionResolver
-                .resolvePartitionDistribution(
-                    brokerConfiguration.getExperimental().getPartitioning(),
-                    brokerConfiguration.getCluster())
-                .partitions());
+        brokerStartupContext.getActorSchedulingService(), staticConfiguration);
   }
 
-  private ClusterTopologyManagerService getClusterTopologyManagerService(
+  private static ClusterTopologyManagerService getClusterTopologyManagerService(
       final BrokerStartupContext brokerStartupContext) {
     final var rootDirectory =
         Path.of(brokerStartupContext.getBrokerConfiguration().getData().getDirectory());
@@ -111,7 +106,7 @@
         );
   }
 
-  private ClusterTopologyGossiperConfig getDefaultClusterTopologyGossipConfig() {
+  private static ClusterTopologyGossiperConfig getDefaultClusterTopologyGossipConfig() {
     return new ClusterTopologyGossiperConfig(Duration.ofSeconds(10), Duration.ofSeconds(2), 2);
   }
 }
Index: broker/src/main/java/io/camunda/zeebe/broker/partitioning/topology/PartitionDistributionResolver.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/broker/src/main/java/io/camunda/zeebe/broker/partitioning/topology/PartitionDistributionResolver.java b/broker/src/main/java/io/camunda/zeebe/broker/partitioning/topology/PartitionDistributionResolver.java
--- a/broker/src/main/java/io/camunda/zeebe/broker/partitioning/topology/PartitionDistributionResolver.java	(revision 6f1ecfadd20bc27f34be1ecfdcd9612cc8392043)
+++ b/broker/src/main/java/io/camunda/zeebe/broker/partitioning/topology/PartitionDistributionResolver.java	(revision 1180e012dfcc8567edf6b6f89d21f46db0c18aae)
@@ -30,7 +30,9 @@
 /** Manage how partitions are distributed among the brokers. */
 public class PartitionDistributionResolver {
 
-  public PartitionDistribution resolvePartitionDistribution(
+  private PartitionDistributionResolver() {}
+
+  public static PartitionDistribution resolvePartitionDistribution(
       final PartitioningCfg partitionCfg, final ClusterCfg clusterCfg) {
     final var partitionDistributor = buildPartitionDistributor(partitionCfg);
     final var clusterMembers = getRaftGroupMembers(clusterCfg);
@@ -79,7 +81,8 @@
         .toList();
   }
 
-  public PartitionDistribution resolvePartitionDistribution(final ClusterTopology clusterTopology) {
+  public static PartitionDistribution resolvePartitionDistribution(
+      final ClusterTopology clusterTopology) {
     if (clusterTopology.isUninitialized()) {
       throw new IllegalStateException(
           "Cannot generated partition distribution from uninitialized topology");
@@ -101,13 +104,14 @@
 
     final var partitionDistribution =
         partitionsToMembersMap.entrySet().stream()
-            .map(this::getPartitionMetadata)
+            .map(PartitionDistributionResolver::getPartitionMetadata)
             .collect(Collectors.toSet());
 
     return new PartitionDistribution(partitionDistribution);
   }
 
-  private PartitionMetadata getPartitionMetadata(final Entry<Integer, Map<MemberId, Integer>> e) {
+  private static PartitionMetadata getPartitionMetadata(
+      final Entry<Integer, Map<MemberId, Integer>> e) {
     final Map<MemberId, Integer> memberPriorities = e.getValue();
     final var optionalPrimary = memberPriorities.entrySet().stream().max(Entry.comparingByValue());
     if (optionalPrimary.isEmpty()) {
@@ -121,7 +125,7 @@
         optionalPrimary.get().getKey());
   }
 
-  private PartitionId partitionId(final Integer key) {
+  private static PartitionId partitionId(final Integer key) {
     return PartitionId.from(PartitionManagerImpl.GROUP_NAME, key);
   }
 }
Index: broker/src/main/java/io/camunda/zeebe/broker/partitioning/topology/StaticClusterTopologyService.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/broker/src/main/java/io/camunda/zeebe/broker/partitioning/topology/StaticClusterTopologyService.java b/broker/src/main/java/io/camunda/zeebe/broker/partitioning/topology/StaticClusterTopologyService.java
--- a/broker/src/main/java/io/camunda/zeebe/broker/partitioning/topology/StaticClusterTopologyService.java	(revision 6f1ecfadd20bc27f34be1ecfdcd9612cc8392043)
+++ b/broker/src/main/java/io/camunda/zeebe/broker/partitioning/topology/StaticClusterTopologyService.java	(revision 1180e012dfcc8567edf6b6f89d21f46db0c18aae)
@@ -24,10 +24,9 @@
   public ActorFuture<Void> start(final BrokerStartupContext brokerStartupContext) {
     try {
       partitionDistribution =
-          new PartitionDistributionResolver()
-              .resolvePartitionDistribution(
-                  brokerStartupContext.getBrokerConfiguration().getExperimental().getPartitioning(),
-                  brokerStartupContext.getBrokerConfiguration().getCluster());
+          PartitionDistributionResolver.resolvePartitionDistribution(
+              brokerStartupContext.getBrokerConfiguration().getExperimental().getPartitioning(),
+              brokerStartupContext.getBrokerConfiguration().getCluster());
     } catch (final Exception e) {
       return CompletableActorFuture.completedExceptionally(e);
     }
Index: broker/src/test/java/io/camunda/zeebe/broker/partitioning/topology/StaticPartitionDistributionResolverTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/broker/src/test/java/io/camunda/zeebe/broker/partitioning/topology/StaticPartitionDistributionResolverTest.java b/broker/src/test/java/io/camunda/zeebe/broker/partitioning/topology/StaticPartitionDistributionResolverTest.java
--- a/broker/src/test/java/io/camunda/zeebe/broker/partitioning/topology/StaticPartitionDistributionResolverTest.java	(revision 6f1ecfadd20bc27f34be1ecfdcd9612cc8392043)
+++ b/broker/src/test/java/io/camunda/zeebe/broker/partitioning/topology/StaticPartitionDistributionResolverTest.java	(revision 1180e012dfcc8567edf6b6f89d21f46db0c18aae)
@@ -56,8 +56,7 @@
 
     // when
     final var topology =
-        new PartitionDistributionResolver()
-            .resolvePartitionDistribution(partitioningCfg, clusterCfg);
+        PartitionDistributionResolver.resolvePartitionDistribution(partitioningCfg, clusterCfg);
     final Set<PartitionMetadata> partitionDistribution = topology.partitions();
 
     // then
@@ -120,8 +119,7 @@
 
     // when
     final var topology =
-        new PartitionDistributionResolver()
-            .resolvePartitionDistribution(partitioningCfg, clusterCfg);
+        PartitionDistributionResolver.resolvePartitionDistribution(partitioningCfg, clusterCfg);
     final Set<PartitionMetadata> partitionDistribution = topology.partitions();
 
     // then
@@ -167,7 +165,7 @@
 
     // when
     final var partitionDistribution =
-        new PartitionDistributionResolver().resolvePartitionDistribution(topology);
+        PartitionDistributionResolver.resolvePartitionDistribution(topology);
 
     // then
     assertThat(partitionDistribution.partitions()).containsExactlyInAnyOrderElementsOf(expected);
@@ -208,7 +206,7 @@
 
     // when
     final var partitionDistribution =
-        new PartitionDistributionResolver().resolvePartitionDistribution(topology);
+        PartitionDistributionResolver.resolvePartitionDistribution(topology);
 
     // then
     assertThat(partitionDistribution.partitions()).containsExactlyInAnyOrderElementsOf(expected);
Index: restore/src/main/java/io/camunda/zeebe/restore/RestoreManager.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/restore/src/main/java/io/camunda/zeebe/restore/RestoreManager.java b/restore/src/main/java/io/camunda/zeebe/restore/RestoreManager.java
--- a/restore/src/main/java/io/camunda/zeebe/restore/RestoreManager.java	(revision 6f1ecfadd20bc27f34be1ecfdcd9612cc8392043)
+++ b/restore/src/main/java/io/camunda/zeebe/restore/RestoreManager.java	(revision 1180e012dfcc8567edf6b6f89d21f46db0c18aae)
@@ -101,9 +101,8 @@
 
     final var factory = new RaftPartitionGroupFactory();
     final var clusterTopology =
-        new PartitionDistributionResolver()
-            .resolvePartitionDistribution(
-                configuration.getExperimental().getPartitioning(), configuration.getCluster());
+        PartitionDistributionResolver.resolvePartitionDistribution(
+            configuration.getExperimental().getPartitioning(), configuration.getCluster());
     // snapshot store factory can be null because we are not going start the partitions.
     final var partitionsGroup =
         factory.buildRaftPartitionGroup(configuration, clusterTopology, null);
Index: topology/src/main/java/io/camunda/zeebe/topology/ClusterTopologyManagerService.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/topology/src/main/java/io/camunda/zeebe/topology/ClusterTopologyManagerService.java b/topology/src/main/java/io/camunda/zeebe/topology/ClusterTopologyManagerService.java
--- a/topology/src/main/java/io/camunda/zeebe/topology/ClusterTopologyManagerService.java	(revision 6f1ecfadd20bc27f34be1ecfdcd9612cc8392043)
+++ b/topology/src/main/java/io/camunda/zeebe/topology/ClusterTopologyManagerService.java	(revision 1180e012dfcc8567edf6b6f89d21f46db0c18aae)
@@ -8,15 +8,14 @@
 package io.camunda.zeebe.topology;
 
 import io.atomix.cluster.ClusterMembershipService;
-import io.atomix.cluster.MemberId;
 import io.atomix.cluster.messaging.ClusterCommunicationService;
-import io.atomix.primitive.partition.PartitionMetadata;
 import io.camunda.zeebe.scheduler.Actor;
 import io.camunda.zeebe.scheduler.ActorSchedulingService;
 import io.camunda.zeebe.scheduler.future.ActorFuture;
 import io.camunda.zeebe.scheduler.future.CompletableActorFuture;
 import io.camunda.zeebe.topology.TopologyInitializer.FileInitializer;
 import io.camunda.zeebe.topology.TopologyInitializer.GossipInitializer;
+import io.camunda.zeebe.topology.TopologyInitializer.StaticConfiguration;
 import io.camunda.zeebe.topology.TopologyInitializer.StaticInitializer;
 import io.camunda.zeebe.topology.TopologyInitializer.SyncInitializer;
 import io.camunda.zeebe.topology.gossip.ClusterTopologyGossiper;
@@ -27,9 +26,6 @@
 import java.io.IOException;
 import java.io.UncheckedIOException;
 import java.nio.file.Path;
-import java.util.List;
-import java.util.Set;
-import java.util.function.Supplier;
 
 public final class ClusterTopologyManagerService extends Actor {
   // Use a node 0 as always the coordinator. Later we can make it configurable or allow changing it
@@ -77,33 +73,30 @@
   }
 
   private TopologyInitializer getCoordinatorInitializer(
-      final Supplier<Set<PartitionMetadata>> staticPartitionResolver,
-      final List<MemberId> knownMembers) {
+      final StaticConfiguration staticConfiguration) {
     return new FileInitializer(persistedClusterTopology)
         .orThen(
             new SyncInitializer(
                 persistedClusterTopology,
-                knownMembers,
+                staticConfiguration.knownMembers(),
                 this,
                 clusterTopologyGossiper::queryClusterTopology))
-        .orThen(new StaticInitializer(staticPartitionResolver, persistedClusterTopology));
+        .orThen(
+            new StaticInitializer(
+                staticConfiguration.partitionDistribution(), persistedClusterTopology));
   }
 
   /** Starts ClusterTopologyManager which initializes ClusterTopology */
   public ActorFuture<Void> start(
       final ActorSchedulingService actorSchedulingService,
-      final List<MemberId> clusterMembers,
-      final Supplier<Set<PartitionMetadata>> partitionDistributionResolver) {
+      final StaticConfiguration staticConfiguration) {
     final var startFuture = new CompletableActorFuture<Void>();
     actorSchedulingService
         .submitActor(this)
         .onComplete(
             (ignore, error) -> {
               if (error == null) {
-                actor.run(
-                    () ->
-                        startClusterTopologyServices(
-                            partitionDistributionResolver, clusterMembers, startFuture));
+                actor.run(() -> startClusterTopologyServices(staticConfiguration, startFuture));
               } else {
                 startFuture.completeExceptionally(error);
               }
@@ -113,12 +106,11 @@
   }
 
   private void startClusterTopologyServices(
-      final Supplier<Set<PartitionMetadata>> partitionDistributionResolver,
-      final List<MemberId> knownMembers,
+      final StaticConfiguration staticConfiguration,
       final CompletableActorFuture<Void> startFuture) {
     final TopologyInitializer topologyInitializer =
         isCoordinator
-            ? getCoordinatorInitializer(partitionDistributionResolver, knownMembers)
+            ? getCoordinatorInitializer(staticConfiguration)
             : getNonCoordinatorInitializer();
     // Start gossiper first so that when ClusterTopologyManager initializes the topology, it can
     // immediately gossip it.
Index: topology/src/main/java/io/camunda/zeebe/topology/TopologyInitializer.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/topology/src/main/java/io/camunda/zeebe/topology/TopologyInitializer.java b/topology/src/main/java/io/camunda/zeebe/topology/TopologyInitializer.java
--- a/topology/src/main/java/io/camunda/zeebe/topology/TopologyInitializer.java	(revision 6f1ecfadd20bc27f34be1ecfdcd9612cc8392043)
+++ b/topology/src/main/java/io/camunda/zeebe/topology/TopologyInitializer.java	(revision 1180e012dfcc8567edf6b6f89d21f46db0c18aae)
@@ -19,13 +19,12 @@
 import io.camunda.zeebe.topology.state.PartitionState;
 import java.io.IOException;
 import java.time.Duration;
-import java.util.List;
 import java.util.Map;
 import java.util.Map.Entry;
+import java.util.Objects;
 import java.util.Set;
 import java.util.function.Consumer;
 import java.util.function.Function;
-import java.util.function.Supplier;
 import java.util.stream.Collectors;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -67,6 +66,18 @@
     };
   }
 
+  /**
+   * Holds statically known configuration of the cluster. This is used to initialize the topology on
+   * the coordinator.
+   */
+  record StaticConfiguration(
+      Set<PartitionMetadata> partitionDistribution, Set<MemberId> knownMembers) {
+    public StaticConfiguration {
+      Objects.requireNonNull(partitionDistribution);
+      Objects.requireNonNull(knownMembers);
+    }
+  }
+
   /** Initialized topology from the locally persisted topology */
   class FileInitializer implements TopologyInitializer {
 
@@ -141,13 +152,13 @@
     private static final Duration SYNC_QUERY_RETRY_DELAY = Duration.ofSeconds(5);
     private final PersistedClusterTopology persistedClusterTopology;
     private final ActorFuture<Boolean> initialized;
-    private final List<MemberId> knownMembersToSync;
+    private final Set<MemberId> knownMembersToSync;
     private final ConcurrencyControl executor;
     private final Function<MemberId, ActorFuture<ClusterTopology>> syncRequester;
 
     public SyncInitializer(
         final PersistedClusterTopology persistedClusterTopology,
-        final List<MemberId> knownMembersToSync,
+        final Set<MemberId> knownMembersToSync,
         final ConcurrencyControl executor,
         final Function<MemberId, ActorFuture<ClusterTopology>> syncRequester) {
       this.persistedClusterTopology = persistedClusterTopology;
@@ -223,20 +234,18 @@
   class StaticInitializer implements TopologyInitializer {
 
     private static final Logger LOGGER = LoggerFactory.getLogger(StaticInitializer.class);
-    private final Supplier<Set<PartitionMetadata>> staticPartitionResolver;
+    private final Set<PartitionMetadata> partitionDistribution;
     private final PersistedClusterTopology persistedClusterTopology;
 
     public StaticInitializer(
-        final Supplier<Set<PartitionMetadata>> staticPartitionResolver,
+        final Set<PartitionMetadata> staticPartitionDistribution,
         final PersistedClusterTopology persistedClusterTopology) {
-      this.staticPartitionResolver = staticPartitionResolver;
+      partitionDistribution = staticPartitionDistribution;
       this.persistedClusterTopology = persistedClusterTopology;
     }
 
     @Override
     public ActorFuture<Boolean> initialize() {
-      final var partitionDistribution = staticPartitionResolver.get();
-
       final var partitionsOwnedByMembers =
           partitionDistribution.stream()
               .flatMap(
Index: topology/src/test/java/io/camunda/zeebe/topology/TopologyInitializerTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/topology/src/test/java/io/camunda/zeebe/topology/TopologyInitializerTest.java b/topology/src/test/java/io/camunda/zeebe/topology/TopologyInitializerTest.java
--- a/topology/src/test/java/io/camunda/zeebe/topology/TopologyInitializerTest.java	(revision 6f1ecfadd20bc27f34be1ecfdcd9612cc8392043)
+++ b/topology/src/test/java/io/camunda/zeebe/topology/TopologyInitializerTest.java	(revision 1180e012dfcc8567edf6b6f89d21f46db0c18aae)
@@ -28,7 +28,6 @@
 import java.nio.file.Path;
 import java.nio.file.StandardOpenOption;
 import java.time.Duration;
-import java.util.List;
 import java.util.Map;
 import java.util.Set;
 import java.util.concurrent.ExecutionException;
@@ -131,7 +130,7 @@
   @Test
   void shouldInitializeFromSync() throws IOException {
     // given
-    final var knownMembers = List.of(MemberId.from("1"));
+    final var knownMembers = Set.of(MemberId.from("1"));
     final ActorFuture<ClusterTopology> syncResponseFuture = new TestActorFuture<>();
     final Function<MemberId, ActorFuture<ClusterTopology>> syncRequester = id -> syncResponseFuture;
     final var initializer =
@@ -156,7 +155,7 @@
   @Test
   void shouldCompleteFutureButNotInitializeWhenSyncReturnsUninitialized() throws IOException {
     // given
-    final var knownMembers = List.of(MemberId.from("1"));
+    final var knownMembers = Set.of(MemberId.from("1"));
     final ActorFuture<ClusterTopology> syncResponseFuture = new TestActorFuture<>();
     final Function<MemberId, ActorFuture<ClusterTopology>> syncRequester = id -> syncResponseFuture;
     final var initializer =
@@ -181,7 +180,7 @@
         Set.of(
             new PartitionMetadata(
                 PartitionId.from("test", 1), Set.of(member), Map.of(member, 1), 1, member));
-    return new StaticInitializer(() -> partitions, persistedClusterTopology);
+    return new StaticInitializer(partitions, persistedClusterTopology);
   }
 
   @Nested
@@ -251,7 +250,7 @@
     @Test
     void shouldInitializeFromSyncWhenFileIsEmpty() {
       // given
-      final var knownMembers = List.of(MemberId.from("1"));
+      final var knownMembers = Set.of(MemberId.from("1"));
       final ActorFuture<ClusterTopology> syncResponseFuture = new TestActorFuture<>();
       final Function<MemberId, ActorFuture<ClusterTopology>> syncRequester =
           id -> syncResponseFuture;
@@ -275,7 +274,7 @@
     @Test
     void shouldInitializeFromStaticWhenFileAndSyncFails() {
       // given
-      final var knownMembers = List.of(MemberId.from("1"));
+      final var knownMembers = Set.of(MemberId.from("1"));
       final ActorFuture<ClusterTopology> syncResponseFuture = new TestActorFuture<>();
       final Function<MemberId, ActorFuture<ClusterTopology>> syncRequester =
           id -> syncResponseFuture;
```
</details>

Reply:
>  It's a minor detail but this supplier is only used in StaticInitializer. IMO it would be easier to understand if we just calculate configured distribution here and pass that down instead of the supplier.

The partition resolver has some sanity checks for the configuration. So if the broker restarts with an incorrect configuration, it might fail the bootstrap even though it is never going to using it. That's why I left it as a supplier.

May be it is better to pass the `PartitionDistributor` instead of the supplier? I didn't do it initially to limit the refactorings. But it makes sense to keep it in the topology module.

Reply:
> The partition resolver has some sanity checks for the configuration. So if the broker restarts with an incorrect configuration, it might fail the bootstrap even though it is never going to using it. That's why I left it as a supplier.

Oh yeah, I haven't thought about that :+1: 

> May be it is better to pass the PartitionDistributor instead of the supplier? 

If that's possible (module dependencies etc.) that'd would be great I think :+1: